stonesoup/predictor/particle.py

```diff
@@ -91,3 +97,137 @@ def predict(self, prior, *args, **kwargs):
+            timestamp=timestamp)
+
+        for model_index, transition_model in enumerate(self.transition_models):
+            # Change the value of the dynamic value randomly according to the defined
+            # transition matrix
+            current_model_indices = prior.dynamic_model == model_index
+            current_model_count = np.count_nonzero(current_model_indices)
+            if current_model_count == 0:
+                continue
+            new_dynamic_models = np.searchsorted(
+                self.probabilities[model_index],
+                np.random.random(size=current_model_count))
+
+            new_state_vector = self.apply_model(
+                prior[current_model_indices], transition_model, timestamp,
+                self.position_mappings[model_index], noise=True, **kwargs)
+
+            new_particle_state.state_vector[:, current_model_indices] = new_state_vector
+            new_particle_state.dynamic_model[current_model_indices] = new_dynamic_models
+
+        return new_particle_state
+
+    @staticmethod
+    def apply_model(prior, transition_model, timestamp, position_mapping, **kwargs):
+        # Based on given position mapping create a new state vector that contains only the
+        # required states
+        orig_ndim = prior.ndim
+        prior.state_vector = prior.state_vector[position_mapping, :]
+        new_state_vector = transition_model.function(
+            prior, time_interval=timestamp - prior.timestamp, **kwargs)
+
+        # Calculate the indices removed from the state vector to become compatible with the
+        # dynamic model
+        for j in range(orig_ndim):
+            if j not in position_mapping:
+                new_state_vector = np.insert(new_state_vector, j, 0, axis=0)
+
+        return new_state_vector
+
+
+class RaoBlackwellisedMultiModelPredictor(MultiModelPredictor):
+    """ParticlePredictor class
+
+    An implementation of a Particle Filter predictor.
+    """
+
+    @predict_lru_cache()
+    def predict(self, prior, timestamp=None, **kwargs):
+        """Particle Filter prediction step
+
+        Parameters
+        ----------
+        prior : :class:`~.RaoBlackwellisedParticleState`
+            A prior state object
+        timestamp: :class:`datetime.datetime`, optional
+            A timestamp signifying when the prediction is performed
+            (the default is `None`)
+        Returns
+        -------
+        : :class:`~.ParticleStatePrediction`
+            The predicted state
+        """
+
+        new_particle_state = Prediction.from_state(
+            prior,
+            state_vector=copy.copy(prior.state_vector),
+            parent=prior,
+            particle_list=None,
+            timestamp=timestamp)
+
+        # Change the value of the dynamic value randomly according to the
+        # matrix
+        new_dynamic_models = np.array([
+            np.random.choice(len(self.transition_models), p=model_probabiliy)
```

Comment:
I came across an issue with the following line of code, causing the error:
"_TypeError: Cannot cast array data from dtype('O') to dtype('float64') according to the rule 'safe'_"
This can be resolved by casting the array as a float, e.g:
```suggestion
            np.random.choice(len(self.transition_models), p=model_probabiliy.astype(float)
```